command_line_tool/platypus.1
.Dd 1/11/05
.Dt platypus 1
.Os Darwin
.Sh NAME
.Nm platypus
.Nd create Mac OS X application wrapper around a script.
.Sh SYNOPSIS
.Nm
.Op Fl vh
.Op Fl FASDBR
.Op Fl P Ar profile
.Op Fl a Ar appName
.Op Fl c Ar scriptPath
.Op Fl o Ar outputType
.Op Fl i Ar icon
.Op Fl p Ar interpreter
.Op Fl V Ar version
.Op Fl s Ar signature
.Op Fl I Ar identifier
.Op Fl f Ar bundledFile
.Op Fl X Ar suffixes
.Op Fl T Ar filetypes
.Op Fl N Ar environment
.Op Fl G Ar arguments
.Ar appName.app
.Sh DESCRIPTION
.Nm
is the command line counterpart of the Platypus Mac OS X application.  It is a command line too 
that creates Mac OS X application bundles that in turn execute a script.
.Pp
.Nm
supports the following flags (executed in the order which they are passed):
.Pp
.Bl -tag -width -Fl
.It Fl p Ar [profile]
Loads all settings from a profile file generated by the Platypus.app Mac OS X application.  You will
still need to specify a destination path for the application.
.It Fl a Ar [appName]
Specifies the name of the application to be created.  This is can be different from the name of the .app
bundle itself, and is displayed in the application's menus, about window and Info.plist property list.
.It Fl c Ar [scriptPath]
Path to the script that is to be wrapped into an application.
.It Fl o Ar [outputType]
Specifies the script application's output type.
Four different kinds of (case sensitive) values are supported as arguments:
.Pp
.Ar 'None'
The application has no graphical output.  Script text output goes to STDOUT.
.Pp
.Ar 'Progress Bar'
The application displays an indeterminate progress bar.  Script output goes to STDOUT.
.Pp
.Ar 'Text Window'
The application displays a window with a text field which receives all script output.
.Pp
.Ar 'Web'
The application displays a window with a web view which renders output as HTML.
.Pp
.It Fl i Ar [icon]
Specifies a file to use as icon for the application bundle.
The icon file can be a regular Mac OS X .icns file, or an image file of any kind
supported by the Mac OS X Cocoa APIs.  As of writing, this includes JPEG, GIF, PNG, 
TIFF, PSD, BMP, PICT, TGA, SGI, PNTG PDF, EPS and PostScript (.ps).  The image will
be scaled to 128x128 pixel size, which is standard for Mac OS X icons.
.Pp
.It Fl p Ar [interpreter]
Sets the interpreter for your script (e.g. /usr/bin/perl).  If no interpreter is specified, the default is the default shell (/bin/sh).
.Pp
.It Fl V Ar [version]
Sets the application bundle's version.  This is reflected in the application bundle's Info.plist
property list and About window.
.Pp
.It Fl u Ar [author]
Sets the name of the application author (e.g. "Apple Computer" or "John Smith").  If not specified, platypus will default to the current user's full user name.
.Pp
.It Fl s Ar [signature]
Mostly for backwards compatibility and heritage reasons.  In Mac OS 9 and earlier (and, indeed, in 
Mac OS X), each application was assigned a four character ASCII signature known as an OSType.  It 
is possible to register application signatures in a database on the Apple Developer Website.   
Please note that Apple has reserved all lowercase-only combinations.  
.Pp
Chances are that you will have no need of this option. 
If it is left empty, the application will get the signature string '????', which is ignored
by Mac OS X.  This value is stored within the application bundle in the Info.plist application 
propertly list.
.Pp
.It Fl f Ar [bundledFile]
Specifies a file to be bundled with the script application.  This file will be copied over the Resources
folder of the application bundle, which is the same folder in which your script will eventually reside.  You
can specify any number of files to be bundled, but you must use the -f flag each time, since
the flag only accepts one file path after the -f flag.
.Pp
.It Fl I Ar [identifier]
Sets the application's bundle identifier.  An application identifier is a kind of reverse DNS name
(e.g. com.apple.iTunes).  If this option is left empty, 
.Nm
will default to an identifier of the format "org.username.appname" (e.g. org.sveinbjornt.Platypus).
.It Fl F
If set, your script will receive the path to the application (e.g. "/Applications/MyPlatypusApp.app") as the first argument ($ARGV[1], $1 etc., depending on your scripting language of choice).
.It Fl A
This flag makes the application request Administrator privileges via Apple's Security Framework (i.e.
prompt for a password) and then execute the script with those privileges.
.Pp
.It Fl S
Secure bundled script.  This encrypts the bundled script, to fend off prying eyes.
.Pp
Please note that this will NOT make the script inaccessible to an
experienced hacker, but just might serve to dissuade the less determined.
.Pp
.It Fl D
Makes your script application "droppable", i.e. capable of receiving drag'n dropped files as arguments
to the script. The application bundle's property list is modified so
that it can receive dropped files in the Dock and Finder. 
These files are then passed on to the script as arguments via @ARGV. 
.Pp
.It Fl B
This option causes the application to run in the background.  The application will not appear in the Dock or bring
its windows to the front when launched.  This is achieved by registering the application with the Finder
as a user interface element (LSUIElement).
.Pp
.It Fl R
This option sets the application so that it remains running after the script has been executed.  It can then later be quit by the user via regular means (i.e. Quit menu item).
.Pp
.It Fl X Ar [suffixes]
Only appropriate if you are using the -D option. This flag allows you to specify the file suffixes
(e.g. .txt, .wav) your application can open.  This should be a |-separated string (e.g. "txt|wav|jpg").
.It Fl T Ar [filetypes]
Only appropriate if you are using the -D option.  This flag allows you to specify the file type codes (e.g. '8BIM', 'TEXT') your application can open.  This should be a |-separated string of 4-character ASCII strings (e.g. "8BIM|TEXT|R*ch").
.It Fl G Ar [arguments]
Arguments for the script interpreter.  These should be specified as a |-separated string (e.g. '-w|-s|-l').
.It Fl N Ar [environemnt]
Sets environmental variables for your script.  These should be specified as a |-separated string of name=value pairs (e.g. 'EDITOR=pico|TERM=xterm-color').
.Pp
.It Fl v
Prints the version of the platypus command line utility
.Pp
.It Fl h
Prints help and usage string
.El
.Pp
The 
.Nm 
utility exits 0 on success, and >0 if an error occurs.
.Pp
.Sh FILES
.Bl -tag -width "//usr/local/share/platypus/PlatypusDefault.icns" -compact
.It Pa /usr/local/bin/platypus
program binary
.It Pa /usr/local/share/platypus/ScriptExec
executable for script app
.It Pa /usr/local/share/platypus/MainMenu.nib
Nib file for script app
.It Pa /usr/local/share/platypus/PlatypusDefault.icns
Default icons
.El
.Sh AUTHORS 
This manual page was written by
.An Sveinbjorn Thordarson Aq sveinbjornt@simnet.is .

command_line_tool/platypus.m
/*
    platypus - command line equivalent to the Mac OS X Platypus application
			 - create application wrappers around scripts
			 
    Copyright (C) 2008 Sveinbjorn Thordarson <sveinbjornt@simnet.is>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*/

/*  CHANGE LOG

	4.0	- Rewritten for Platypus 4.0 release
	1.3	- File bundling, suffixes/types, for 3.4 release
	1.2 - For 3.3 release
    1.0 - * First release of the Platypus command line tool
*/

/*
	SUPPORT FILES FOR PROGRAM (defined in CommonDefs.h)

	/usr/local/share/platypus/exec						Executable bundled with app
	/usr/local/share/platypus/MainMenu.nib				Nib file for app
	/usr/local/share/platypus/PlatypusDefault.icns		Default icons for Platypus apps

*/

/*
	COMMAND LINE OPTIONS
	
	// *required*
	-a	Application Name
	-o	Output Type
	-c	Script File
	-P  Profile
	
	// *advanced options*
	
	-i	Icon file
	-p	Interpreter
	-V	Version
	-s	Signature (4 character string)
	-I	Identifier
	-f  Bundled file argument

	-A	Requires Administrator privileges
	-S	Secure bundled script
	-D	Is Droppable
	-B	Runs in background
	-R  remainRunningAfterCompletion
	
	-X  Suffixes supported
	-T  File Type Codes supported
	
	-b  Verbose output
	-v	Print version string
	-h	Print help
*/

///////////// IMPORTS/INCLUDES ////////////////

#import <Cocoa/Cocoa.h>
#import <AppKit/AppKit.h>

#import "CommonDefs.h"
#import "PlatypusAppSpec.h"

#include <stdio.h>
#include <unistd.h>
#include <errno.h>
#include <sys/stat.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>

///////////// DEFINITIONS ////////////////

#define		OPT_STRING			"P:c:f:a:t:o:i:u:p:V:s:I:ASDFBRvhX:T:G:N:b" 

//#define		DEBUG				1

///////////// PROTOTYPES ////////////////

static void PrintVersion (void);
static void PrintUsage (void);
static void PrintHelp (void);
static void NSPrintErr (NSString *str);
static void NSPrint (NSString *str);


int main (int argc, const char * argv[]) 
{
    NSAutoreleasePool	*pool				= [[NSAutoreleasePool alloc] init];//set up autorelease pool
	NSApplication		*app				= [NSApplication sharedApplication];//establish connection to Window Server
	NSFileManager		*fm					= [NSFileManager defaultManager];
	
	// we start with an application spec set to all the default settings
	// command line params can fill in the settings user wants
	PlatypusAppSpec		*appSpec			= [[PlatypusAppSpec alloc] initWithDefaults];

	int					rc, optch;
    static char			optstring[] = OPT_STRING;

	NSPrint(@"Starting argument parsing");

    while ( (optch = getopt(argc, (char * const *)argv, optstring)) != -1)
    {
        switch(optch)
        {
			// Profile
			case 'P':
			{
				NSString *profilePath = [NSString stringWithCString: optarg];
				if (![fm fileExistsAtPath: profilePath])
				{
					NSPrintErr(@"Profile path invalid\n");
					exit(1);
				}
				[appSpec release];
				appSpec = [[PlatypusAppSpec alloc] initWithProfile: profilePath];
			}
			break;
		
			// App Name
			case 'a':				
				[appSpec setProperty: [NSString stringWithCString: optarg] forKey: @"Name"];
				break;
			
			// A bundled file
			case 'f':
			{
				NSString *filePath = [[NSString stringWithCString: optarg] stringByExpandingTildeInPath];
				if (![fm fileExistsAtPath: filePath])
				{
					NSPrintErr([NSString stringWithFormat: @"No file exists at path '%@'\n", filePath]);
					exit(1);
				}
				[[appSpec propertyForKey: @"BundledFiles"] addObject: filePath];
			}
			break;
			
			// Script path
			case 'c':
			{
				NSString *scriptPath = [[NSString stringWithCString: optarg] stringByExpandingTildeInPath];
				if (![fm fileExistsAtPath: scriptPath])
				{
					NSPrintErr([NSString stringWithFormat: @"No script file exists at path '%@'\n", scriptPath]);
					exit(1);
				}
				[appSpec setProperty: scriptPath forKey: @"ScriptPath"];
			}
			break;
		
			// Output Type
            case 'o':
			{
				NSString *outputType = [NSString stringWithCString: optarg];
				if ([outputType caseInsensitiveCompare: @"None"] != NSOrderedSame &&
					[outputType caseInsensitiveCompare: @"Progress Bar"] != NSOrderedSame &&
					[outputType caseInsensitiveCompare: @"Text Window"] != NSOrderedSame &&
					[outputType caseInsensitiveCompare: @"Web"] != NSOrderedSame)
				{
						NSPrintErr(@"Invalid output type.\n");
						exit(1);
				}
				[appSpec setProperty: [NSString stringWithCString: optarg] forKey: @"Output"];
			}
			break;
			
			// Author
			case 'u':
				[appSpec setProperty: [NSString stringWithCString: optarg] forKey: @"Author"];
				break;
			
			// Icon
			case 'i':
			{
				NSString *iconPath = [[NSString stringWithCString: optarg] stringByExpandingTildeInPath];
				if (![fm fileExistsAtPath: iconPath])
				{
					NSPrintErr([NSString stringWithFormat: @"No icon file exists at path '%@'\n", iconPath]);
					exit(1);
				}
				
				 // specifying an icns file, that means we just use the file
				if ([iconPath hasSuffix: @"icns"])
				{
					[appSpec setProperty: [NSNumber numberWithBool: YES] forKey: @"HasCustomIcon"];
					[appSpec setProperty: [NSString stringWithCString: optarg] forKey: @"CustomIconPath"];
				}
				else
				{
					// read image from file
					NSImage *iconImage = [[NSImage alloc] initWithContentsOfFile: iconPath];
					if (iconImage == NULL)
					{
						NSPrintErr([NSString stringWithFormat: @"Unable to get icon from file '%@'\n", iconPath]);
						exit(1);
					}
					[appSpec setProperty: [NSNumber numberWithBool: NO] forKey: @"HasCustomIcon"];
					[appSpec setProperty: [NSKeyedArchiver archivedDataWithRootObject: iconImage] forKey: @"Icon"];
					[iconImage release];
				}
			}
			break;
			
			// Interpreter
			case 'p':
				[appSpec setProperty: [NSString stringWithCString: optarg] forKey: @"Interpreter"];
				break;
			
			// Version
			case 'V':
				[appSpec setProperty: [NSString stringWithCString: optarg] forKey: @"Version"];
				break;
			
			// Signature
			case 's':
			{
				NSString *signatureStr = [NSString stringWithCString: optarg];
				if ([signatureStr length] != 4) // it must be a 4-character string
				{
					NSPrintErr(@"Signature invalid.\nA signature must consist of exactly 4 ASCII characters.\n");
					exit(1);
				}
				[appSpec setProperty: signatureStr forKey: @"Signature"];
			}
			break;
			
			// The checkbox options
			case 'I':
				[appSpec setProperty: [NSString stringWithCString: optarg] forKey: @"Identifier"];
				break;
            case 'A':
				[appSpec setProperty: [NSNumber numberWithBool: YES] forKey: @"Authentication"];
				break;
			case 'S':
				[appSpec setProperty: [NSNumber numberWithBool: YES] forKey: @"Secure"];
				break;
			case 'D':
				[appSpec setProperty: [NSNumber numberWithBool: YES] forKey: @"Droppable"];
				break;
			case 'F':
				[appSpec setProperty: [NSNumber numberWithBool: YES] forKey: @"AppPathAsFirstArg"];
				break;
			case 'B':
				[appSpec setProperty: [NSNumber numberWithBool: YES] forKey: @"ShowInDock"];				
				break;
			case 'R':
				[appSpec setProperty: [NSNumber numberWithBool: YES] forKey: @"RemainRunning"];
				break;
				
			// Suffixes
			case 'X':
			{
				NSString *suffixesStr = [NSString stringWithCString: optarg];
				NSArray *suffixes = [suffixesStr componentsSeparatedByString: @"|"];
				[appSpec setProperty: suffixes forKey:  @"Suffixes"];
			}
			break;
			
			// File Types
			case 'T':
			{
				NSString *filetypesStr = [NSString stringWithCString: optarg];
				NSArray *fileTypes = [filetypesStr componentsSeparatedByString: @"|"];
				[appSpec setProperty: fileTypes forKey: @"FileTypes"];
			}
			break;
			
			// Parameters for interpreter
			case 'G':
			{
				NSString *parametersString = [NSString stringWithCString: optarg];
				NSArray *parametersArray = [parametersString componentsSeparatedByString: @"|"];
				[appSpec setProperty: parametersArray forKey: @"Parameters"];
			}
			break;
			
			// Environment
			case 'N':
			{
				NSMutableDictionary *envDict = [[[NSMutableDictionary alloc] initWithCapacity: 255] autorelease];
				int i;
				NSString *environmentVarsString = [NSString stringWithCString: optarg];
				NSArray *envVarDefs = [environmentVarsString componentsSeparatedByString: @"|"];
				for (i = 0; i < [envVarDefs count]; i++)
				{
					NSArray *nameValPair = [[envVarDefs objectAtIndex: i] componentsSeparatedByString: @"="];
					[envDict setObject: [nameValPair objectAtIndex: 1] forKey: [nameValPair objectAtIndex: 0]];
				}
				[appSpec setProperty: envDict forKey: @"Environment"];
			}
			break;
			
			case 'v':
                PrintVersion();
                exit(0);
                break;
            case 'h':
                PrintHelp();
                return 0;
                break;
            default:
                rc = 1;
                PrintUsage();
                return 0;
        }
    }
	
	if (argc - optind < 1) //  application destination must follow
    {
        NSPrintErr(@"Too few arguments.\n");
        PrintUsage();
        exit(1);
    }
			
	//get application destination parameter
	NSString *appPath = [[NSString stringWithCString: argv[optind]] stringByExpandingTildeInPath];
	if (appPath == NULL)
	{
		NSPrintErr(@"Missing parameter:  Application Destination Path\n");
		PrintUsage();
		exit(1);
	}
	
	[appSpec setProperty: appPath forKey: @"Destination"];
		
	// set nib and exec paths
	[appSpec setProperty: CMDLINE_NIB_PATH forKey: @"NibPath"];
	[appSpec setProperty: CMDLINE_EXEC_PATH forKey: @"ExecutablePath"];
	
	// create the app from spec
	if (![appSpec verify] || ![appSpec create])
	{
		NSPrintErr([appSpec getError]);
		return 1;
	}
	
	[appSpec release];
    [pool release];
	
	return 0;
}

#pragma mark -

////////////////////////////////////////
// Print version and author to stdout
///////////////////////////////////////

static void PrintVersion (void)
{
    NSPrint([NSString stringWithFormat: @"%@ version %@ by %@\n", CMDLINE_PROGNAME, PROGRAM_VERSION, PROGRAM_AUTHOR]);
}

////////////////////////////////////////
// Print usage string to stdout
///////////////////////////////////////

static void PrintUsage (void)
{
    NSPrint([NSString stringWithFormat: @"usage: %@ [-vh] [-P profile] [-a name] [-o outputType] [-p interpreter] [-FASDBRiuVsIXTGN] script destination\n", CMDLINE_PROGNAME]);
}

////////////////////////////////////////
// Print help string to stdout
///////////////////////////////////////

static void PrintHelp (void)
{
	puts("platypus - command line application wrapper generator for scripts");
	PrintVersion();
    PrintUsage();
	puts("");
	puts("Options:");
	puts("");
	puts("-P [profile]		Load settings from profile file");
	puts("-a [name]		Set name of application bundle");
	puts("-o [type]		Set output type.  See man page for accepted types");
	puts("-c [script]		Set script for application");
	puts("-p [interpreter]	Set interpreter for script");
	puts("");
	puts("-i [icon]		Set icon for application");
	puts("-u [author]		Set name of application author");
	puts("-V [version]		Set version of application");
	puts("-s [signature]		Set 4-character bundle signature");
	puts("-I [identifier]		Set bundle identifier (i.e. org.yourname.appname)");
	puts("");
	puts("-F			Script receives path to app as first argument");
	puts("-A			App runs with Administrator privileges");
	puts("-S			Secure bundled script");
	puts("-D			App accepts dropped files as argument to script");
	puts("-B			App runs in background (LSUI Element)");
	puts("-R			App remains running after executing script");
	puts("");
	puts("-X [suffixes]		Set suffixes handled by application");
	puts("-T [filetypes]		Set file type codes handled by application");
	puts("-G [arguments]		Set arguments for script interpreter");
	puts("-N [environment]	Set environmental variables for script");
	puts("");
	puts("-h			Prints help");
	puts("-v			Prints program name, version and author");
	puts("");
}

#pragma mark -

/**************************************************
void NSPrint (NSString *str)
Utility function -- dumps NSString to stdout
************************************************/
static void NSPrint (NSString *str)
{
     [str writeToFile: @"/dev/stdout" atomically: NO];
}

static void NSPrintErr (NSString *str)
{
	[str writeToFile: @"/dev/stderr" atomically: NO];
}
